llvm/lib/CodeGen/ReplaceWithVeclib.cpp

```diff
@@ -108,8 +108,22 @@ static bool replaceWithCallToVeclib(const TargetLibraryInfo &TLI,
   // all vector operands match the previously found EC.
   SmallVector<Type *, 8> ScalarArgTypes;
   Intrinsic::ID IID = II->getIntrinsicID();
+
+  // OloadTys collects types used in scalar intrinsic overload name.
+  SmallVector<Type *, 3> OloadTys;
+  if (VTy && isVectorIntrinsicWithOverloadTypeAtArg(IID, -1))
+    OloadTys.push_back(VTy->getElementType());
+
   for (auto Arg : enumerate(II->args())) {
     auto *ArgTy = Arg.value()->getType();
+    // Gather type if it is used in the overload name.
+    if (isVectorIntrinsicWithOverloadTypeAtArg(IID, Arg.index())) {
```

Comment:
are there any tests, testing this logic? I think the pow case is working thanks to the condition added to line 114

Reply:
The `pow` case is impacted by both this logic and return type logic, since `OloadTys` should include only the return type for `pow`, not each of the argument types.  `isVectorIntrinsicWithOverloadTypeAtArg()` is how you determine the arguments that the intrinsic is overloaded on, when not `void`, and defaults to -1 (return type).

Other tests will be running through this logic (and turning out the same), but the pow one is the only case with two arguments which had existing tests meant to match a vector lib call right now.  `atan2` will be added soon, which also needs this.  Other intrinsics could require this change in theory, if there's a matching vector lib call for them.

Reply:
There are three tests impacted by this change.  If you think there's a specific deficiency at this point not addressed by existing testing, please let me know.  For now, I'll be resolving this comment.  Please re-activate if you think it still needs addressing.

Reply:
ah, there is one unit tests using "powi", which is overloaded on a scalar type.
For now we do not have mappings for it in the TLI, so there is no lit tests for it.


The regular pow is not affecting this code path as is only overloaded only on return type.